Tutorial for cross-compilation (#160)
nix-dot-dev/nix.dev: source/tutorials/cross-compilation.rst

.. _ref-cross-compilation:

Cross-compilation
=================

Cross-compilation is the act of **compiling code** on the **build platform**

> to the **host platform**, where the compiled **executable runs**. [#]_

Reviewer: Those platforms also need to differ.

Author: You mean it's not possible for build platform to also be the host platform?

Reviewer: That is possible, but that is native compilation, usually. However, you can also do “native” cross compilation (like exherbo) which is also [a goal for nixpkgs I subscribe to](https://github.com/NixOS/nixpkgs/issues/21471).

Since cross compilation is the more generic concept, it is probably hard disambiguate — in a way native compilation is a special case of cross compilation. Following that train of thought you could talk about cross compilation as soon as your compiler has a notion of a target platform.

I think my original suggestion may not be the best one, but the initial paragraph reads to me as if it is missing something, maybe?

Author: How about a second paragraph that says:

> Cross-compilation is generalized from native compilation, where host and build platforms are the same.

Reviewer: Why not!

It's needed when the host platform has limited resources (such as CPU)
or when it's not easily accessible for development.

Nix community has world-class support for cross-compilation,

> after years of hard work from our community.

Reviewer: Style: community is repeated here a lot which doesn't read to nice. Also I feel like “The Nix community” sounds more correct, but unsure.

.. [#] Terminology for cross-compilation platforms differs between build systems,
       Nix community has chosen to follow 
       `autoconf terminology <https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Hosts-and-Cross_002dCompilation.html>`_.

.. note:: macOS/Darwin is a special case, as not the whole OS is Open Source. 

>           It's only possible to cross-compile between ``aarch64-darwin`` and ``x86_64-darwin``.

Reviewer: Technically only from `x86_64-darwin` to `aarch64-darwin`. Not sure if I'd mention this here though, as `aarch64-darwin` is far from mature at this point and cross compilation is not working at all in any channel  currently…

Author: > Technically only from x86_64-darwin to aarch64-darwin.

Is that because we're missing `pkgsCross.x86_64-darwin` or something else?

> Not sure if I'd mention this here though, as aarch64-darwin is far from mature at this point and cross compilation is not working at all in any channel currently…

Since the fix from @alyssais should work, the whole cross-compilation is not well supported but I'd like this tutorial to show that it's entirely feasible with enough maintainers. WDYT?

Reviewer: > Is that because we're missing pkgsCross.x86_64-darwin or something else?

I'm not sure as to why exactly, but I can confirm that nixpkgs can't even cross compile from x86_64-darwin to x86_64-darwin due to an evaluation problem very early in `stdenv`.

> the whole cross-compilation is not well supported but I'd like this tutorial to show that it's entirely feasible with enough maintainers

Still, aarch64-linux for example has had more time to mature, accumulate fixes etc. Maybe not mentioning aarch64-darwin saves someone a bit of frustration?

Author: How about a disclaimer that we've just added support for it?

Author: I've added a disclaimer and move this further down.

What's a target platform?
-------------------------

There's actually a third platform named target.

It matters in cases where you'd like to distribute a compiler binary, 
as you'd then like to build a compiler on the build platform, compile code on the

> target plaform and run the final executable on the host platform.

Reviewer: This is wrong: You would run the compiler on the *host* platform and it would generate code for the *target* platform.

This is also why targetPlatform usually doesn't matter: host platform means the same for everything and for everything but build tools and compilers targetPlatform makes zero difference .

> Since that's rarely needed, we'll treat target platform the same as the build.

Reviewer: Just write that you ignore the platform for this tutorial. The default case of targetPlatform is the same as hostPlatform afaik.

Also interesting to note: On the outermost interface nixpkgs doesn't support `targetPlatform`, it only accepts `localSystem` (~ build) and `crossSystem` (~ host) since cross compiling a cross compiler is a rare use case outside of stdenv bootstrapping.

Author: I'd like to clarify since it took me quite some time to figure out why target is not important and explain why it can be ignored to save confusion. The user will eventually stumble upon it as they dive into cross-compilation.

Reviewer: I've thought about this yesterday and maybe it would actually *help* the tutorial to explain target completely. This would make it very easy to clear the following things up:

* Why are binutils, pkg-config and cc prefixed? What does the prefix mean? This is quite easy to answer if you explain the relation of the target prefix to the target platform
* What is `buildPackages`? Here having an understanding of target is nice: `buildPackages` usually has build and host platform equivalent to `localSystem`, but its target is `crossSystem` — having an understanding of target automatically explains how we bootstrap our cross package sets

Author: Agreed! I don't think I know enough to really explain those, unfortunately.

Determining the host platform
-----------------------------

The build platform is determined automatically by Nix

> as it can just guess it during the configure phase.

Reviewer: You can also set it by passing `localSystem`.

Author: In what cases is that really needed? In most cases guessing should be sufficient?

Reviewer: The two main use cases would be remote build (if you have a builder with a different platform than your local machine) and testing evaluation of cross-related things by simulating being on another platform.

Author: I'd like to mention that then in #157 tutorial (and consider developing nixpkgs out of scope here).

The host platform is best determined by running on the host platform:

.. code:: shell-session 

  $ bash $(nix-build '<nixpkgs>' -A gnu-config)/config.guess
  aarch64-unknown-linux-gnu

In case that's not possible (when the host platform is not easily accessible
for development), it has to be constructed manually via the following template:

.. code::

  <cpu>-<vendor>-<os>-<abi>

Note that ``<vendor>`` is often ``unknown`` and ``<abi>`` is optional. 
There's also no unique identifier for a platform, for example ``unknown`` and 
``pc`` are interchangeable (hence it's called config.guess).

Some other common examples of platforms:

- aarch64-apple-darwin14
- aarch64-pc-linux-gnu
- x86_64-w64-mingw32
- aarch64-apple-ios

> 

Reviewer: This section feels superfluous if you never mention the `crossSystem` argument which can actually be used to pass a `config` such as we have gotten in this section to use as the host platform for nixpkgs.

Mentioning `crossSystem` is important because it is the most basic case, i. e. `pkgsCross` is just a convenient shortcut for (re-)importing nixpkgs with `crossSystem` set.

Author: How often do we imagine the user provides their own crossSystem? Currently I want to teach that the single truth is the platform identifier and how it one can go between nixpkgs attributes and the identifier.

In case the user needs a new crossSystem, there's a note how to add one to nixpkgs.

Can you think of a use case where that's not enough?

Reviewer: Well the immediate use case would of course be (as it has come up here) that you can pass a triple to `crossSystem` and it will figure things out (or try to).

Author: Good point :laughing:

  pkgsCross.armhf-embedded              pkgsCross.ppc-embedded
  pkgsCross.armv7a-android-prebuilt     pkgsCross.ppcle-embedded
  pkgsCross.armv7l-hf-multiplatform     pkgsCross.raspberryPi
  pkgsCross.avr                         pkgsCross.remarkable1
  pkgsCross.ben-nanonote                pkgsCross.remarkable2
  pkgsCross.fuloongminipc               pkgsCross.riscv32
  pkgsCross.ghcjs                       pkgsCross.riscv32-embedded
  pkgsCross.gnu32                       pkgsCross.riscv64
  pkgsCross.gnu64                       pkgsCross.riscv64-embedded
  pkgsCross.i686-embedded               pkgsCross.scaleway-c1
  pkgsCross.iphone32                    pkgsCross.sheevaplug
  pkgsCross.iphone32-simulator          pkgsCross.vc4
  pkgsCross.iphone64                    pkgsCross.wasi32
  pkgsCross.iphone64-simulator          pkgsCross.x86_64-embedded
  pkgsCross.mingw32                     pkgsCross.x86_64-netbsd
  pkgsCross.mingwW64                    pkgsCross.x86_64-netbsd-llvm
  pkgsCross.mmix                        pkgsCross.x86_64-unknown-redox
  pkgsCross.msp430                      

From the attribute name it can't always be immediately clear what is the platform.

> It's possible to query the platform name using::

Reviewer: maybe we should stick to terminology of config instead of name in the tutorial?

Real world cross-compiling of a Hello World example
---------------------------------------------------
 
To show off the power of cross-compilation in Nix, let's build our own Hello World program 
by cross-compiling it as static executables to ``armv6l-unknown-linux-gnueabihf``
and ``x86_64-w64-mingw32`` (Windows) platforms and run the resulting executable
with `an emulator <https://en.wikipedia.org/wiki/Emulator>`_.

.. code:: nix 

  { pkgs ? import <nixpkgs> {} 
  }:

  let
    # Create a C program that prints Hello World
    helloWorld = pkgs.writeText "hello.c" ''
      #include <stdio.h>
      
      int main (void)
      {
        printf ("Hello, world!\n");
        return 0;
      }
    '';

    # A function that takes host platform packages
    crossCompileFor = hostPkgs:
      # Run a simple command with the compiler available
      hostPkgs.runCommandCC "hello-world-cross-test" {} ''
        # Wine requires home directory
        HOME=$PWD

        # Compile our example using the compiler specific to our host platform
        $CC ${helloWorld} -o hello 

        # Run the compiled program using an emulator
        # Usually Qemu, but on windows it is Wine

>         ${hostPkgs.stdenv.hostPlatform.emulator hostPkgs.buildPackages} hello > $out

Reviewer: Especially `buildPackages` and maybe how the emulator attribute works should be explained.

        # print to stdout program stdout
        cat $out
      '';
  in {
    # Statically compile our example using the two platform hosts
    rpi = crossCompileFor pkgs.pkgsCross.raspberryPi;
    windows = crossCompileFor pkgs.pkgsCross.mingwW64;
  }

If we build this example and print both resulting derivations, we should see "Hello, world!" for each: 

.. code:: shell-session

  $ cat $(nix-build cross-compile.nix)
  Hello, world!
  Hello, world!

Developer environment with a cross-compiler
-------------------------------------------

In :ref:`tutorial for declarative reproducible environments <declarative-reproducible-envs>`,
we've looked at how Nix helps us provide tooling and system libraries for our project.

It's also possible to provide an environment with a compiler configured for cross-compilation!

Given we have a ``shell.nix``:

.. code:: nix

  { nixpkgs ? fetchTarball "https://github.com/NixOS/nixpkgs/archive/bba3474a5798b5a3a87e10102d1a55f19ec3fca5.tar.gz"
  , pkgs ? (import nixpkgs {}).pkgsCross.aarch64-multiplatform
  }:

  # pkgs.callPackage is needed due to https://github.com/NixOS/nixpkgs/pull/126844

>   pkgs.callPackage ({ mkShell, zlib, pkg-config }: mkShell {

Reviewer: @k-bx probably this is the issue you're encountering.

```suggestion
  pkgs.callPackage ({ mkShell, zlib, file, pkg-config }: mkShell {
```

    # these tools run on the build platform, but are configure to target the target platform
    nativeBuildInputs = [ pkg-config file ];
    # libraries needed for the target platform
    buildInputs = [ zlib ];
  }) {}

And ``hello.c``:

.. code:: c 

  #include <stdio.h>

  int main (void)
  {
    printf ("Hello, world!\n");
    return 0;
  }

We can cross-compile it:

.. code:: shell-session 

  $ nix-shell --run '$CC hello.c -o hello' cross-compile-shell.nix

And confirm it's aarch64:

.. code:: shell-session 

  $ nix-shell --run 'file hello' cross-compile-shell.nix 
  hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /nix/store/733hzlw1hixdm6dfdsb8dlwa2h8fl5qi-glibc-2.31-74-aarch64-unknown-linux-gnu/lib/ld-linux-aarch64.so.1, for GNU/Linux 2.6.32, with debug_info, not stripped
  

Next steps
----------

> - `Official binary cache <https://cache.nixos.org>`_ doesn't come with binaries

Reviewer: This is technically not true, we have some cache for certain things as part of the `cross-trunk` jobset although it should probably be expanded one of these days…

Author: I wanted to avoid that to set expectations low and simplify explanation. Maybe something like `comes with almost no binaries`?

Reviewer: How about something like “We have extremely limited binary-cache for cross compilation, so it is very possible that you have to compile the respective toolchain yourself”? This would also clarify the risk (compiling binutils and gcc).

.. _ref-cross-compilation:

Cross-compilation
=================

> Cross-compilation is the act of **compiling code** on the **build platform**
> to the **host platform**, where the compiled **executable runs**. [#]_

Reviewer 2: ```suggestion
When compiling code, we can distinguish between the **build platform**, where the executable
is *built*, and the **host platform**, where the compiled executable *runs*. [#]_
**Native compilation** is the special case where those two platforms are the same.
**Cross compilation** is the general case where those two platforms are not.
```

Also I wouldn't hyphenate "cross-compilation". It's just a noun phrase. Only "cross-compiled" the adjective as in "let's use this cross-compiled library" needs it.

Author: I took the inspiration from nixpkgs manual: https://nixos.org/manual/nixpkgs/stable/#chap-cross

It's needed when the host platform has limited resources (such as CPU)
or when it's not easily accessible for development.

The Nix community has world-class support for cross-compilation,
after years of hard work from our community.

.. [#] Terminology for cross-compilation platforms differs between build systems.
       We have chosen to follow 
       `autoconf terminology <https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Hosts-and-Cross_002dCompilation.html>`_.

> .. note:: macOS/Darwin is a special case, as not the whole OS is open-source. 

Reviewer 3: I think for added clarity, it would be nice to mention it's a special case to use as a build platform.

Author: It's it a special case for the host platform?

If macOS is a build platform, you can still cross compile to linux as a target for example.

Reviewer 2: Yeah macos -> true FOSS is fine. I might also move this minor caviat down, as none of the text above talks about what platforms are supported yet or anything like that.

Reviewer 3: Oh I was confused here, wouldn't it also be the case with windows then? As windows is also closed source largely?

Reviewer 2: It's possible to make macOS cross work better; it's just no one did yet. With WIndows it has been done.

The build platform is determined automatically by Nix
as it can just guess it during the configure phase.

The host platform is best determined by running on the host platform:

.. code:: shell-session 

  $ bash $(nix-build '<nixpkgs>' -A gnu-config)/config.guess
  aarch64-unknown-linux-gnu

In case that's not possible (when the host platform is not easily accessible
for development), the platform config has to be constructed manually via the following template:

.. code::

  <cpu>-<vendor>-<os>-<abi>

Note that ``<vendor>`` is often ``unknown`` and ``<abi>`` is optional. 
There's also no unique identifier for a platform, for example ``unknown`` and 
``pc`` are interchangeable (hence it's called config.guess).

If you can't install Nix, find a way to run ``config.guess`` (usually comes with
 the autoconf package) from the OS you're able to run on the host platform.

Some other common examples of platform configs:

- aarch64-apple-darwin14
- aarch64-pc-linux-gnu
- x86_64-w64-mingw32
- aarch64-apple-ios

Choosing the host platform with Nix
-----------------------------------

Nixpkgs comes with a set of predefined host platforms applied to all packages.

> It's possible to list predefined attribute sets via shell completion:

Reviewer 3: Shell completion doesn't always work fully, so it might be nice to give the command to list these as well. I'm often scratching my head when thinking about how to search for things in nixpkgs so it's good to be reminded often I feel

Author: Good point! I should maybe show how to use a reply here.

Reviewer: `lib/systems/examples.nix` is not a scary file I think it is usable as a list of platforms.

Author: I'd like to avoid pointing the user to a file (through github) as a reference to what attributes `pkgsCross` has as it's hard to explain the connection between the two :(

Reviewer: I mean the alternative is telling them to run

```console
$ nix-instantiate --eval --strict -E 'builtins.attrNames (import <nixpkgs/lib>).systems.examples'
```

pretty much…

Author: Went with `nix repl` as that's a good thing to teach anyway (although it adds more things to introduce here, it's for the better).

  pkgsCross.armv7l-hf-multiplatform     pkgsCross.raspberryPi
  pkgsCross.avr                         pkgsCross.remarkable1
  pkgsCross.ben-nanonote                pkgsCross.remarkable2
  pkgsCross.fuloongminipc               pkgsCross.riscv32
  pkgsCross.ghcjs                       pkgsCross.riscv32-embedded
  pkgsCross.gnu32                       pkgsCross.riscv64
  pkgsCross.gnu64                       pkgsCross.riscv64-embedded
  pkgsCross.i686-embedded               pkgsCross.scaleway-c1
  pkgsCross.iphone32                    pkgsCross.sheevaplug
  pkgsCross.iphone32-simulator          pkgsCross.vc4
  pkgsCross.iphone64                    pkgsCross.wasi32
  pkgsCross.iphone64-simulator          pkgsCross.x86_64-embedded
  pkgsCross.mingw32                     pkgsCross.x86_64-netbsd
  pkgsCross.mingwW64                    pkgsCross.x86_64-netbsd-llvm
  pkgsCross.mmix                        pkgsCross.x86_64-unknown-redox
  pkgsCross.msp430                      

Cross compilation package attribute names are made up, so it isn't always clear 
what is the corresponding platform config.

It's possible to query the platform config using:

>   $ nix-instantiate '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.hostPlatform.config --eval

Reviewer 2: ```suggestion
  $ nix-instantiate '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.stdenv.hostPlatform.config --eval
```
We would like to deprecate `hostPlatform` at the top level.
